scripts/upstreamCommit.sh

```diff
@@ -1,38 +1,41 @@
 #!/usr/bin/env bash
-(
-set -e
-PS1="$"
+
+set -euo pipefail
 
 function changelog() {
-    base=$(git ls-tree HEAD $1  | cut -d' ' -f3 | cut -f1)
-    cd $1 && git log --oneline ${base}...HEAD
+  local -r submodule_dir="${1?Missing submodule dir}"
+  local base
+  base="$(git ls-tree HEAD "${submodule_dir}"  | awk '{ print $3 }')"
+  git -C "${submodule_dir}" log --oneline "${base}"...HEAD
 }
-bukkit=$(changelog work/Bukkit)
-cb=$(changelog work/CraftBukkit)
-spigot=$(changelog work/Spigot)
-
-updated=""
-logsuffix=""
-if [ ! -z "$bukkit" ]; then
-    logsuffix="$logsuffix\n\nBukkit Changes:\n$bukkit"
-    updated="Bukkit"
-fi
-if [ ! -z "$cb" ]; then
-    logsuffix="$logsuffix\n\nCraftBukkit Changes:\n$cb"
-    if [ -z "$updated" ]; then updated="CraftBukkit"; else updated="$updated/CraftBukkit"; fi
-fi
-if [ ! -z "$spigot" ]; then
-    logsuffix="$logsuffix\n\nSpigot Changes:\n$spigot"
-    if [ -z "$updated" ]; then updated="Spigot"; else updated="$updated/Spigot"; fi
-fi
-disclaimer="Upstream has released updates that appear to apply and compile correctly.\nThis update has not been tested by PaperMC and as with ANY update, please do your own testing"
 
-if [ ! -z "$1" ]; then
-    disclaimer="$@"
-fi
+function main() {
+  local disclaimer="${*:-}"
 
-log="${UP_LOG_PREFIX}Updated Upstream ($updated)\n\n${disclaimer}${logsuffix}"
+  if [ -z "${disclaimer:-}" ]; then
+    disclaimer="Upstream has released updates that appear to apply and compile correctly.\nThis update has not been tested by PaperMC and as with ANY update, please do your own testing"
+  fi
 
-echo -e "$log" | git commit -F -
+  local submodule updated logsuffix changes
+  for submodule in Bukkit CraftBukkit Spigot; do
+    changes="$(changelog work/"${submodule}")"
+    if [ -n "${changes:-}" ]; then
+      if [ -z "${updated:-}" ]; then
+        updated="${submodule}"
+      else
+        updated="${updated}/${submodule}"
+      fi
+      logsuffix="${logsuffix:-}\n\n${submodule} Changes:\n${changes}"
+    fi
+  done
 
-) || exit 1
+  if [ -n "${updated:-}" ]; then
+    log="${UP_LOG_PREFIX:-}Updated Upstream ($updated)\n\n${disclaimer}${logsuffix}"
+    echo -e "${log}" | git commit -F -
```

Comment:
`echo -e` is not available on whatever macOS uses I believe

Reply:
I removed it, but then added back because it was used before anyway. Can replace with `printf` builtin.

Reply:
It's kind of up to cat to decide, honestly. I don't have a mac to test.

Reply:
Replaced with printf builtin. Using `%b` format specifier (available in Bash, Ksh and BB) that will make sure to only expand escaped characters and not try to parse printf format specifiers.
See: <https://github.com/koalaman/shellcheck/wiki/SC2059>